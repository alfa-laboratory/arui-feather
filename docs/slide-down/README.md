# SlideDown

Компонент "расхлопа".
Позволяет скрывать и отображать контент.

```javascript
import SlideDown from 'arui-feather/slide-down';
```




## Props


| Prop  | Тип  | По-умолчанию | Обязательный | Описание |
| ----- | ---- | ------------ | ------------ |----------|
| isExpanded | Boolean |  |  | Управление состоянием expand/collapse компонента |
| children | Node |  |  | Контент компонента |
| theme | [ThemeEnum](#ThemeEnum) |  |  | Тема компонента |
| className | String\|Function |  |  | Дополнительный класс |







## Типы






### <a id="ThemeEnum"></a>ThemeEnum

 * `'alfa-on-color'`
 * `'alfa-on-white'`



