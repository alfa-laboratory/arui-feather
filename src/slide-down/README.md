```jsx
import Button from 'arui-feather/button';
import Paragraph from 'arui-feather/paragraph';

function handleSlideDownToggle() {
    setState({ isExpanded: !state.isExpanded });
}
initialState = {
    isExpanded: false
};
<div>
    <Button
        onClick={ handleSlideDownToggle }
    >
        Как узнать дату и сумму платежа
    </Button>
    <div className='row' >
        <SlideDown isExpanded={ state.isExpanded }>
            <Paragraph>
                Узнать сумму платежа по кредиту с ежемесячным погашением равными частями вы
                можете из графика погашения, предоставленного при оформлении кредита.
            </Paragraph>
        </SlideDown>
    </div>
</div>
```
