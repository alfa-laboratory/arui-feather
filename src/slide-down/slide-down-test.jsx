/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { render, cleanUp } from '../test-utils';

import SlideDown from './slide-down';

describe('slide-down', () => {
    afterEach(cleanUp);

    it('should render without problems', () => {
        let { node: slideDownNode } = render(<SlideDown>slide-down-text</SlideDown>);

        expect(slideDownNode).to.exist;
        expect(slideDownNode).to.have.text('slide-down-text');
        expect(slideDownNode).to.have.class('slide-down');
    });

    it('should render with height=0 by default', () => {
        let { node: slideDownNode } = render(
            <SlideDown>
                <div style={ { height: '100px' } } />
            </SlideDown>
        );

        expect(slideDownNode).to.have.attribute('style', 'height: 0px;');
    });

    it('should render with height=auto when isExpanded=true', () => {
        let { node: slideDownNode } = render(
            <SlideDown isExpanded={ true } />
        );

        expect(slideDownNode).to.have.attribute('style', 'height: auto;');
    });

    it('should set `slide-down__content_expanded` class to content when isExpanded=true', () => {
        let { node: slideDownNode } = render(
            <SlideDown isExpanded={ true } />
        );

        let slideDownContentNode = slideDownNode.querySelector('.slide-down__content');
        expect(slideDownContentNode).to.have.class('slide-down__content_expanded');
    });
});
