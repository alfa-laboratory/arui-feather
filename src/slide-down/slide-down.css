/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.slide-down {
    overflow: hidden;
    transition: height .3s ease;
    position: relative;

    &__content {
        position: absolute;
        bottom: 0;
        left: 0;
        right: 0;

        &_expanded {
            position: static;
        }
    }
}
