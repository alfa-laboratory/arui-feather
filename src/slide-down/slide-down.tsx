/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React, { createRef } from 'react';
import { DeepReadonly } from 'utility-types';
import { createCn } from 'bem-react-classname';
import { withTheme } from '../cn';

export type SlideDownProps = DeepReadonly<{

    /**
     * Управление состоянием expand/collapse компонента
     */
    isExpanded?: boolean;

    /**
     * Контент компонента
     */
    children?: React.ReactNode;

    /**
     * Тема компонента
     */
    theme?: 'alfa-on-color' | 'alfa-on-white';

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Идентификатор компонента в DOM
     */
    id?: string;

    /**
     * Обработчик события начала анимации
     */
    onAnimationStart?: () => void;

    /**
     * Обработчик события окончания анимации
     */
    onAnimationEnd?: () => void;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    'data-test-id'?: string;

}>;

type SlideDownState = {
    height: number | string;
    isHeightAuto: boolean;
}

/**
 * Компонент "расхлопа".
 * Позволяет скрывать и отображать контент.
 */
export class SlideDown extends React.PureComponent<SlideDownProps, SlideDownState> {
    protected cn = createCn('slide-down');

    state = {
        height: this.props.isExpanded ? 'auto' : 0,
        isHeightAuto: this.props.isExpanded,
    };

    private slideDown = createRef<HTMLDivElement>();

    private slideDownContent = createRef<HTMLDivElement>()

    componentDidUpdate(prevProps: SlideDownProps) {
        const { isExpanded, onAnimationStart } = this.props;

        if (prevProps.isExpanded !== isExpanded) {
            if (isExpanded) {
                this.setHeightToContentHeight();
            } else {
                this.setHeightToNull();
            }

            if (onAnimationStart) {
                onAnimationStart();
            }
        }
    }

    render() {
        return (
            <div
                className={ this.cn() }
                id={ this.props.id }
                style={
                    { height: this.getHeight() }
                }
                onTransitionEnd={ this.handleTransitionEnd }
                ref={ this.slideDown }
                data-test-id={ this.props['data-test-id'] }
            >
                <div
                    className={ this.cn('content', { expanded: this.state.isHeightAuto }) }
                    ref={ this.slideDownContent }
                >
                    { this.props.children }
                </div>
            </div>
        );
    }

    private handleTransitionEnd = (event: React.TransitionEvent<HTMLDivElement>) => {
        if (event.propertyName === 'height' && this.props.isExpanded) {
            this.setAutoHeight();
        }
        if (this.props.onAnimationEnd) {
            this.props.onAnimationEnd();
        }
    };

    private getHeight() {
        return this.state.isHeightAuto
            ? 'auto'
            : this.state.height;
    }

    private setHeightToContentHeight() {
        if (this.slideDownContent.current) {
            this.setState({
                isHeightAuto: false,
                height: this.slideDownContent.current.offsetHeight,
            });
        }
    }

    private setHeightToNull() {
        this.setHeightToContentHeight();

        // Заставляем React перерисовать элемент
        this.forceUpdate(() => {
            // Заставляем браузер сделать reflow
            if (this.slideDown.current) {
                this.slideDown.current.getBoundingClientRect();
            }

            this.setState({ height: 0 });
        });
    }

    private setAutoHeight() {
        this.setState({
            isHeightAuto: true,
        });
    }
}

class ThemedSlideDown extends SlideDown {}
(ThemedSlideDown as any) = withTheme(SlideDown);
export default ThemedSlideDown;
