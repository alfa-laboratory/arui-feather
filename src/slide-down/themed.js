import { withTheme } from '../cn';
import Component from './slide-down';

export default withTheme(Component);
