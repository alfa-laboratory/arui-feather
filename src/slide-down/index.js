import './slide-down.css';

export default from './slide-down';
