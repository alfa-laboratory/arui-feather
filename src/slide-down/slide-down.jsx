/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React from 'react';
import Type from 'prop-types';

import { autobind } from 'core-decorators';

import cn from '../cn';
import performance from '../performance';

import './slide-down.css';

/**
 * Компонент "расхлопа".
 * Позволяет скрывать и отображать контент.
 */
@cn('slide-down')
@performance()
class SlideDown extends React.Component {
    static propTypes = {
        /** Управление состоянием expand/collapse компонента */
        isExpanded: Type.bool,
        /** Контент компонента */
        children: Type.node,
        /** Тема компонента */
        theme: Type.oneOf(['alfa-on-color', 'alfa-on-white']),
        /** Дополнительный класс */
        className: Type.oneOfType([Type.string, Type.func])
    };

    state = {
        height: 0,
        isHeightAuto: false
    };

    slideDown;
    slideDownContent;

    componentDidMount() {
        if (this.props.isExpanded) {
            this.setAutoHeight();
        }
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.isExpanded !== nextProps.isExpanded) {
            if (nextProps.isExpanded) {
                this.setHeightToContentHeight();
            } else {
                this.setHeightToNull();
            }
        }
    }

    render(cn) {
        return (
            <div
                className={ cn }
                style={
                    { height: this.getHeight() }
                }
                onTransitionEnd={ this.handleTransitionEnd }
                ref={ slideDown => (this.slideDown = slideDown) }
            >
                <div
                    className={ cn('content', { expanded: this.state.isHeightAuto }) }
                    ref={ slideDownContent => (this.slideDownContent = slideDownContent) }
                >
                    { this.props.children }
                </div>
            </div>
        );
    }

    @autobind
    handleTransitionEnd(event) {
        if (event.propertyName === 'height' && this.props.isExpanded) {
            this.setAutoHeight();
        }
    }

    getHeight() {
        return this.state.isHeightAuto
            ? 'auto'
            : this.state.height;
    }

    setHeightToContentHeight() {
        this.setState({
            isHeightAuto: false,
            height: this.slideDownContent.offsetHeight
        });
    }

    setHeightToNull() {
        this.setHeightToContentHeight();

         // Заставляем React перерисовать элемент
        this.forceUpdate(() => {
            // Заставляем браузер сделать reflow
            this.slideDown.getBoundingClientRect();
            this.setState({
                height: 0
            });
        });
    }

    setAutoHeight() {
        this.setState({
            isHeightAuto: true
        });
    }
}

export default SlideDown;
