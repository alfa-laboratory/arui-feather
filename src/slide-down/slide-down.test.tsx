/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React from 'react';
import { shallow, mount } from 'enzyme';

import { SlideDown } from './slide-down';

describe('slide-down', () => {
    it('should render without problems', () => {
        const slideDownNode = shallow(<SlideDown>slide-down-text</SlideDown>);

        expect(slideDownNode).toMatchSnapshot();
        expect(slideDownNode.text()).toContain('slide-down-text');
    });

    it('should render with height=0 by default', () => {
        const slideDownNode = shallow(
            <SlideDown>
                <div style={ { height: '100px' } } />
            </SlideDown>,
        );

        expect(slideDownNode.props().style).toEqual({ height: 0 });
    });

    it('should render with height=auto when isExpanded=true', () => {
        const slideDownNode = shallow(
            <SlideDown isExpanded={ true } />,
        );

        expect(slideDownNode.props().style).toEqual({ height: 'auto' });
    });

    it('should set `slide-down__content_expanded` class to content when isExpanded=true', () => {
        const slideDownNode = shallow(
            <SlideDown isExpanded={ true } />,
        );

        const slideDownContentNode = slideDownNode.find('.slide-down__content');

        expect(slideDownContentNode.props().className).toContain('slide-down__content_expanded');
    });

    it('should call onAnimationStart when received new isExpanded prop', () => {
        const onAnimationStart = jest.fn();
        const props = { isExpanded: false, onAnimationStart };
        const slideDownNode = mount<SlideDown>(
            <SlideDown { ...props } />,
        );

        slideDownNode.setProps({ isExpanded: true });

        expect(onAnimationStart).toHaveBeenCalled();
    });

    it('should call onAnimationEnd when handleTransitionEnd is called', () => {
        const onAnimationEnd = jest.fn();
        const slideDownNode = shallow<SlideDown>(
            <SlideDown isExpanded={ false } onAnimationEnd={ onAnimationEnd } />,
        );

        (slideDownNode.instance() as any).handleTransitionEnd({});
        expect(onAnimationEnd).toHaveBeenCalled();
    });
});
