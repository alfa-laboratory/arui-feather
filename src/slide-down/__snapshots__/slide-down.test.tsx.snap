// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slide-down should render without problems 1`] = `
<div
  className="slide-down"
  onTransitionEnd={[Function]}
  style={
    Object {
      "height": 0,
    }
  }
>
  <div
    className="slide-down__content"
  >
    slide-down-text
  </div>
</div>
`;
